ambry-frontend/src/main/java/com/github/ambry/frontend/PostAccountsHandler.java

```diff
@@ -220,5 +225,51 @@ private void updateAccounts(RetainingAsyncWritableChannel channel) throws RestSe
             RestServiceErrorCode.getRestServiceErrorCode(ase.getErrorCode()));
       }
     }
+
+    /**
+     * Process the request json and call {@link AccountService#addDataset} to add the dataset.
+     * @param channel The {@link RetainingAsyncWritableChannel} to read the bytes out
+     * @throws RestServiceException
+     */
+    private void addDataset(RetainingAsyncWritableChannel channel) throws RestServiceException {
+      Dataset datasetToUpdate;
+      String accountName = null;
+      String containerName = null;
+      String datasetName = null;
+      try {
+        datasetToUpdate = AccountCollectionSerde.datasetsFromInputStreamInJson(channel.consumeContentAsInputStream());
+        if (datasetToUpdate == null) {
+          throw new RestServiceException("The dataset is null after deserialize from given InputStream",
+              RestServiceErrorCode.BadRequest);
+        }
+        datasetName = datasetToUpdate.getDatasetName();
+        String accountNameFromRequest =
+            RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_ACCOUNT_NAME, false);
+        String containerNameFromRequest =
+            RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_CONTAINER_NAME, false);
+        accountName = datasetToUpdate.getAccountName();
+        containerName = datasetToUpdate.getContainerName();
+        if (accountNameFromRequest != null && !accountNameFromRequest.equals(accountName)) {
+          throw new RestServiceException(
+              "There is a mismatch between account name from header: " + accountNameFromRequest
+                  + " and account name from dataset: " + accountName, RestServiceErrorCode.BadRequest);
+        }
+        if (containerNameFromRequest != null && !containerNameFromRequest.equals(containerName)) {
+          throw new RestServiceException(
+              "There is a mismatch between account name from header: " + containerNameFromRequest
+                  + " and account name from dataset: " + containerName, RestServiceErrorCode.BadRequest);
+        }
```

Comment:
this is strange. do we need to check the account name and the container name from the header? Why not just rely on the post body?

Reply:
This is just for some check when people provide the header which does not match with what they provide from dataset. Usually, they will only provide through the dataset. But in some cases, they might accidentally provide the accountName through header.